### PR TITLE
Message Relay & Resource Contracts

### PR DESCRIPTION
**Pull request by martindale**
Adding work for Resource Contracts, including relay policies.

**Included:**
- Moves Bitcoin service tests to `tests/bitcoin`

**TODO:**
- [ ] #34
- [ ] #35

### CODE REVIEW

File: types/reader.js [hunk 1]
```diff
@@ -56,23 +72,52 @@ class Reader extends EventEmitter {
...
+    parts.push(header.slice(0, 4)); // magic
+    parts.push(header.slice(4, 8)); // version
+    parts.push(header.slice(8, 12)); // type
+    parts.push(header.slice(12, 16)); // payload size
+
+    const map = parts.map((x) => {
+      return Buffer.from(x, 'hex');
+    });
+
+    const elements = map.map((x) => {
```
Comment (melnx) [inline types/reader.js:89]: don't need parenthesis around single parameter

File: types/service.js [hunk 2]
```diff
@@ -684,16 +708,19 @@ class Service extends Scribe {
...
    * @return {Promise}       Resolves upon successful registration.
    */
-  async _registerActor (actor) {
-    if (!actor.id) return this.error('Client must have an id.');
-
-    this.emit('message', `Registering Actor: ${actor.id} ${JSON.stringify(actor).slice(0, 32)}…`);
+  async _registerActor (actor = {}) {
+    if (!actor.id) {
+      const entity = new Actor(actor);
+      actor = merge({
```
Comment (melnx) [inline types/service.js:714]: could use ... operator here

File: types/service.js [hunk 3]
```diff
@@ -702,54 +729,79 @@ class Service extends Scribe {
...
+    this.emit('actor', registration);
+
+    return registration;
   }
 
   async _registerChannel (channel) {
-    if (!channel.id) return this.error('Channel must have an id.');
+    if (!channel.id) {
+      const entity = new Actor(channel);
+      channel = merge({
```
Comment (melnx) [inline types/service.js:742]: try to replace merge usages with ... operator

File: types/cli.js [hunk 4]
```diff
@@ -481,7 +484,9 @@ class CLI extends App {
   }
 
   async _handlePeerDebug (message) {
-    this._appendDebug(message);
+    if (this.settings.debug) {
```
Comment (melnx) [inline types/cli.js:487]: redundant to check if debug is on because _appendDebug does it already

**Review by melnx — CHANGES_REQUESTED** [review]
comments reflect request for changes in multiple places, not just where the comment is

File: types/service.js (hunk 2 above)
Comment (martindale) [reply types/service.js:714]: Most of these should recurse down the property tree, you'll see this pattern elsewhere

**Comment by martindale** [comment]
Added @jfabric as this contains architectural changes and overview — will return here after upcoming demo

**Pull request merged**